src/main/scala/cromwell/engine/CallActor.scala

```diff
@@ -1,46 +1,84 @@
 package cromwell.engine
 
+import java.util.UUID
+
 import akka.actor.{Actor, Props}
 import akka.event.{Logging, LoggingReceive}
 import cromwell.binding._
 import cromwell.binding.values.WdlValue
-import cromwell.engine
-import cromwell.engine.backend.Backend
+import cromwell.engine.backend.{TaskAbortedException, Backend}
 import cromwell.engine.workflow.WorkflowActor
 import cromwell.engine.workflow.WorkflowActor.CallFailed
 
 import scala.language.postfixOps
-import scala.util.{Failure, Success}
-
+import scala.util.{Try, Failure, Success}
 
 object CallActor {
+
   sealed trait CallActorMessage
-  case class Start(inputs: CallInputs) extends CallActorMessage
-  case class ExecutionFinished(outputs: CallOutputs) extends CallActorMessage
-  case class ExecutionFailed(regret: Throwable) extends CallActorMessage
+  case class Start(call: Call, locallyQualifiedInputs: Map[String, WdlValue], backend: Backend, workflowDescriptor: WorkflowDescriptor) extends CallActorMessage
+  case class RegisterCallAbortFunction(abortFunction: AbortFunction) extends CallActorMessage
+  case object AbortCall extends CallActorMessage
+  case class ExecutionFinished(call: Call, outputs: Try[CallOutputs]) extends CallActorMessage
 
-  def props(call: Call, locallyQualifiedInputs: Map[String, WdlValue], backend: Backend, workflowDescriptor: WorkflowDescriptor): Props =
-    Props(new CallActor(call, locallyQualifiedInputs, backend, workflowDescriptor))
+  def props(callReference: CallReference): Props = Props(new CallActor(callReference))
 }
 
 /** Actor to manage the execution of a single call. */
-class CallActor(call: Call, locallyQualifiedInputs: Map[String, WdlValue], backend: Backend, workflowDescriptor: WorkflowDescriptor) extends Actor with CromwellActor {
+class CallActor(callReference: CallReference) extends Actor with CromwellActor {
+
+  import CallActor._
 
   type CallOutputs = Map[String, WdlValue]
+  
+  var abortFunction: Option[AbortFunction] = None
+  var abortRequested = false
```

Comment:
The more I look at this the more I think this is crying out for `FSM` (or at least `become()`) but I prefer the former. There are multiple discrete states - NotStarted, Started, Aborting (and perhaps things like Finished and Aborted).

It wouldn't be much in the way of less code (it might even be slightly more) but there'd be less branching and keying off of mutable state, IMO a net-win for comprehension